README.md
# react-key-handler 🔑

[![npm version](https://img.shields.io/npm/v/react-key-handler.svg)](https://www.npmjs.com/package/react-key-handler) [![License](https://img.shields.io/npm/l/react-key-handler.svg)](https://www.npmjs.com/package/react-key-handler) [![Build Status](https://travis-ci.org/ayrton/react-key-handler.svg?branch=master)](https://travis-ci.org/ayrton/react-key-handler)

React component to handle keyboard events (such as `keyup`, `keydown` & `keypress`).

## Testimonials

> <div>“Happy to see that react-key-handler is SSR safe :+1:”</div>
> [Veselin Todorov](https://github.com/vesln), Chai.js core

## Table of Contents

1. [Installation](#installation)
2. [Usage](#usage)
   1. [Higher-order Components](#higher-order-components)
   2. [Component](#component)
   3. [Form key handling](#form-key-handling)
3. [Key event names](#key-event-names)
4. [`keyValue`, `code` and `keyCode`](#keyvalue-code-and-keycode)
5. [Development](#development)
   1. [Setup](#setup)
   2. [Getting started](#getting-started)
   3. [Tests](#tests)
6. [Contributing](#contributing)
7. [License](#license)

## Installation

```sh
$ npm install react-key-handler --save
```

## Usage

You can use `react-key-handler` library in two flavours:

- [higher-order components](#higher-order-components)
- [component](#component)

### Higher-order Components

This library includes two similar higher-order components, but with a different puprose:

| Higher-order Component | Purpose             |
| ---------------------- | ------------------- |
| `keyHandler`           | Handles key changes |
| `keyToggleHandler`     | Handles key toggles |

Both have the same API and will decorate the given component with a `keyValue`, `code`
and `keyCode` property.

Internally the `KeyHandler` component is used, for a full understanding be sure to
check out [the implementation](lib/key-handler.js).

```jsx
import React from 'react';
import { keyHandler, KEYPRESS } from 'react-key-handler';

function Demo({ keyValue }) {
  return (
    <div>
      {keyValue === 's' && (
        <ol>
          <li>hello</li>
          <li>world</li>
        </ol>
      )}
    </div>
  );
}

export default keyHandler({ keyEventName: KEYPRESS, keyValue: 's' })(Demo);
```

The prop types of the `KeyHandler` component are:

| Name            | Type   | Required   | Default   |                                        |
| --------------- | ------ | ---------- | --------- | -------------------------------------- |
| keyEventName    | string | no         | `'keyup'` | `'keydown'`, `'keypress'` or `'keyup'` |
| keyValue        | string | yes **\*** |           | Any given [KeyboardEvent.key]          |
| code            | string | yes **\*** |           | Any given [KeyboardEvent.code]         |
| keyCode&dagger; | number | yes **\*** |           | Any given [KeyboardEvent.keyCode]      |

**\*** You should pass at least one of these props.

&dagger; _Note_ that the keyCode is frequently browser specific and has therefore be set as
deprecated, see [MDN for details](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/keyCode).

[Examples](demo/components/examples/decorators/)

### Component

```jsx
import React from 'react';
import KeyHandler, { KEYPRESS } from 'react-key-handler';

export default class Demo extends React.Component {
  state = { showMenu: false };

  render() {
    const { showMenu } = this.state;

    return (
      <React.Fragment>
        <KeyHandler
          keyEventName={KEYPRESS}
          keyValue="s"
          onKeyHandle={this.toggleMenu}
        />

        {showMenu && (
          <ol>
            <li>hello</li>
            <li>world</li>
          </ol>
        )}
      </React.Fragment>
    );
  },

  toggleMenu = (event) => {
    event.preventDefault();

    this.setState({ showMenu: !this.state.showMenu });
  };
}
```

The prop types of the `KeyHandler` component are:

| Name            | Type     | Required   | Default   |                                                  |
| --------------- | -------- | ---------- | --------- | ------------------------------------------------ |
| keyEventName    | string   | no         | `'keyup'` | `'keydown'`, `'keypress'` or `'keyup'`           |
| keyValue        | string   | yes **\*** |           | Any given [KeyboardEvent.key]                    |
| code            | string   | yes **\*** |           | Any given [KeyboardEvent.code]                   |
| keyCode&dagger; | number   | yes **\*** |           | Any given [KeyboardEvent.keyCode]                |
| onKeyHandle     | function | yes        |           | Function that is called when they key is handled |

**\*** You should pass at least one of these props.

&dagger; _Note_ that the keyCode is frequently browser specific and has therefore be set as
deprecated, see [MDN for details](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/keyCode).

[Example](demo/components/examples/component/index.js)

### Form key handling

This library does not handle key events for form elements such as `<input />` and `<textarea />`.

React does a fine job supporting these already via [keyboard events](https://reactjs.org/docs/events.html#keyboard-events).

[Examples](demo/components/examples/input/)

## Key event names

TODO: explain the differences between the different key events.

## `keyValue`, `code` and `keyCode`

The three available key events are

- `keyValue` This corresponds to the true value. This is the value of the key pressed
  by the user while taking into considerations the state of modifier keys
  such as the `shiftKey` as well as the keyboard locale/layout
- `code` This corresponds to the physical key on the keyboard (as opposed to the
  character generated by pressing the key). In other words, this property
  returns a value which isn't altered by keyboard layout or the state of
  the modifier keys. The value is a string specific to the key, e.g. 'Digit0'
- `keyCode` This is similar to code but numeric and also _deprecated_.

We recommend you to use the new Web standard [KeyboardEvent.key] or the [KeyboardEvent.code]
over the deprecated [KeyboardEvent.keyCode].

Note that in **React** `key` is a reserved property, and thus we use `keyValue` when referring
to the `key` property.

**Browser support:**

There's no need to worry about browser support because internally we normalize
deprecated HTML5 `keyValue` values and translate from legacy `keyCode` values,
similar to how React does this for their `SyntheticKeyboardEvent`.

**More information:**

[W3C Working Draft].

## Development

### Setup

```sh
$ git clone <this repo>
$ cd react-key-handler
$ npm install
```

### Getting started

To start the server:

```sh
$ npm demo
```

This starts a development server, which will automatically rebuild the demo app as you change files and supports hot module replacement for fast development:

```sh
$ open http://localhost:1234
```

### Tests

To run all tests:

```sh
$ npm test
```

Or you can run the linters, unit tests and check for type errors individually:

```sh
$ npm run test:lint
$ npm run test:unit
$ npm run test:flow
```

## Contributing

Bug reports and pull requests are welcome on GitHub. This project is intended to be a
safe, welcoming space for collaboration, and contributors are expected to adhere
to the [Contributor Covenant](http://contributor-covenant.org/) code of conduct.

## License

```
 _________________
< The MIT License >
 -----------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||
```

[w3c working draft]: https://www.w3.org/TR/DOM-Level-3-Events-key/
[keyboardevent.key]: https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/key
[keyboardevent.code]: https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/code
[keyboardevent.keycode]: https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/keyCode
[key]: https://facebook.github.io/react/docs/create-fragment.html


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]

- [bugfix]
- [feature]

## [1.2.0-beta.3] - 2018-08-30

- [feature] Add `code` support by [@gforge](https://github.com/gforge)
- [deprecate] `keyCode` prop in favour of `code` by [@gforge](https://github.com/gforge)
- [feature] Allows for `keyValue`, `code` and `keyCode` to come in array format by [@gforge](https://github.com/gforge)
- [breaking] No longer add a `<div />` wrapper around the `KeyHandler` component, but use `React.Fragment` instead

## [1.2.0-beta.2] - 2018-08-21

- [bugfix] Pass original props to component we are decorating when using one of the decorators

## [1.2.0-beta.1] - 2018-08-21

- [feature] Use rollup to bundle up the code

## [1.1.0] - 2018-07-31

> There's two major bugs in 1.1.0, it is highly recommended to update to 1.2.0 or downgrade to 1.0.1

- [feature] Add flow support by [@gforge](https://github.com/gforge)
- [bugfix] Fallback from `event.keyCode` to `event.which` for browsers such as FireFox when using the `keypress` event name. [@gforge](https://github.com/gforge)

## [1.0.1] - 2017-10-26

- [feature] Loosen react dependency to support React 16

## [1.0.0] - 2017-08-28

- [breaking] Drop keyName support
- [bugfix] Get rid of deprecation warnings

## [0.3.0] - 2016-09-13

- [feature] Treat content editable elements as inputs by [@jamesfzhang](https://github.com/jamesfzhang). This means the key handle
  will be ignored for events triggered from these elements. (Draft.js support)

## [0.2.0] - 2016-04-17

- [feature] Loosen react dependency for future react versions
- [deprecate] `keyName` prop in favour of `keyValue`
- [feature] Added prop types
- [feature] Add support for W3C `KeyboardEvent.key` by [@leocavalcante](https://github.com/leocavalcante)

## [0.1.0] - 2016-03-02

- [feature] Add decorators
- [feature] Ignore key events from form elements
- [feature] Key names

## [0.0.4] - 2016-02-27

- [bugfix] Use right function to remove the event listener

## [0.0.2] & [0.0.3] - 2016-02-27

- [bugfix] Protect from server side rendering errors

## [0.0.1] - 2016-02-27

- Initial implementation

[unreleased]: https://github.com/ayrton/react-key-handler/compare/v1.1.0...HEAD
[1.2.0-beta.3]: https://github.com/ayrton/react-key-handler/compare/v1.2.0-beta.2...v1.2.0-beta.3
[1.2.0-beta.2]: https://github.com/ayrton/react-key-handler/compare/v1.2.0-beta.1...v1.2.0-beta.2
[1.2.0-beta.1]: https://github.com/ayrton/react-key-handler/compare/v1.1.0...v1.2.0-beta.1
[1.1.0]: https://github.com/ayrton/react-key-handler/compare/v1.0.1...v1.1.0
[1.0.1]: https://github.com/ayrton/react-key-handler/compare/v1.0.0...v1.0.1
[1.0.0]: https://github.com/ayrton/react-key-handler/compare/v0.3.0...v1.0.0
[0.3.0]: https://github.com/ayrton/react-key-handler/compare/v0.2.0...v0.3.0
[0.2.0]: https://github.com/ayrton/react-key-handler/compare/v0.1.0...v0.2.0
[0.1.0]: https://github.com/ayrton/react-key-handler/compare/v0.0.4...v0.1.0
[0.0.4]: https://github.com/ayrton/react-key-handler/compare/v0.0.3...v0.0.4
[0.0.3]: https://github.com/ayrton/react-key-handler/compare/v0.0.2...v0.0.3
[0.0.2]: https://github.com/ayrton/react-key-handler/compare/v0.0.1...v0.0.2
[0.0.1]: https://github.com/ayrton/react-key-handler/commit/8267e3dc7357bb7fb106f5148e6f9cb9f69ed3b5


lib/constants.js
/* @flow */
/**
 * Key event names.
 */

export const KEYDOWN = 'keydown';
export const KEYPRESS = 'keypress';
export const KEYUP = 'keyup';

export type KeyConstants = 'keydown' | 'keypress' | 'keyup';

export const NORMALIZED_KEYS: {
  [key: string]: string,
} = {
  Esc: 'Escape',
  Spacebar: ' ',
  Left: 'ArrowLeft',
  Up: 'ArrowUp',
  Right: 'ArrowRight',
  Down: 'ArrowDown',
  Del: 'Delete',
  Win: 'OS',
  Menu: 'ContextMenu',
  Apps: 'ContextMenu',
  Scroll: 'ScrollLock',
  MozPrintableKey: 'Unidentified',
};

export const KEY_CODE_KEYS: {
  [string: string]: string,
} = {
  '8': 'Backspace',
  '9': 'Tab',
  '12': 'Clear',
  '13': 'Enter',
  '16': 'Shift',
  '17': 'Control',
  '18': 'Alt',
  '19': 'Pause',
  '20': 'CapsLock',
  '27': 'Escape',
  '32': ' ',
  '33': 'PageUp',
  '34': 'PageDown',
  '35': 'End',
  '36': 'Home',
  '37': 'ArrowLeft',
  '38': 'ArrowUp',
  '39': 'ArrowRight',
  '40': 'ArrowDown',
  '45': 'Insert',
  '46': 'Delete',
  '112': 'F1',
  '113': 'F2',
  '114': 'F3',
  '115': 'F4',
  '116': 'F5',
  '117': 'F6',
  '118': 'F7',
  '119': 'F8',
  '120': 'F9',
  '121': 'F10',
  '122': 'F11',
  '123': 'F12',
  '144': 'NumLock',
  '145': 'ScrollLock',
  '224': 'Meta',
};
